src/connector/src/sink/mongodb.rs

```diff
@@ -405,49 +386,51 @@ impl MongodbSinkWriter {
                 }
                 continue;
             }
-            self.payload_writer.append(insert_builder, row).await?;
+            self.payload_writer.append(insert_builder, row)?;
         }
         Ok(())
     }
 
-    async fn upsert(&mut self, chunk: StreamChunk) -> Result<()> {
+    fn upsert(&mut self, chunk: StreamChunk) -> Result<()> {
         let upsert_builder =
             must_match!(&mut self.command_builder, CommandBuilder::Upsert(builder) => builder);
         for (op, row) in chunk.rows() {
             if op == Op::UpdateDelete {
                 // we should ignore the `UpdateDelete` in upsert mode
                 continue;
             }
-            self.payload_writer.upsert(upsert_builder, op, row).await?;
+            self.payload_writer.upsert(upsert_builder, op, row)?;
         }
         Ok(())
     }
 }
 
-#[async_trait]
-impl SinkWriter for MongodbSinkWriter {
-    async fn begin_epoch(&mut self, _epoch: u64) -> Result<()> {
-        Ok(())
-    }
+pub type MongodbSinkDeliveryFuture = impl TryFuture<Ok = (), Error = SinkError> + Unpin + 'static;
 
-    async fn write_batch(&mut self, chunk: StreamChunk) -> Result<()> {
+impl AsyncTruncateSinkWriter for MongodbSinkWriter {
+    type DeliveryFuture = MongodbSinkDeliveryFuture;
+
+    async fn write_chunk<'a>(
+        &'a mut self,
+        chunk: StreamChunk,
+        mut add_future: DeliveryFutureManagerAddFuture<'a, Self::DeliveryFuture>,
+    ) -> Result<()> {
         if self.is_append_only {
-            self.append(chunk).await
+            self.append(chunk)?;
```

Comment:
It seems that we will always drain the map of `CommandBuilder` after we write to the command builder. If so, instead of making `command_builder` a field of the writer, we can remove the field, and change to return a new `CommandBuilder` in `append` and `upsert`. And then we can have a `into_futures` method for the `CommandBuilder` to turn it into futures to be await on.